src/SubscriptionManager.ts

```diff
@@ -2,30 +2,55 @@ import { NetworkNode } from 'streamr-network'
 import { Todo } from './types'
 
 export class SubscriptionManager {
-
+    publishSessionTimeout: number
     networkNode: NetworkNode
     streams: Map<Todo,Todo>
+    publishTimeouts: Map<string, NodeJS.Timeout>
 
-    constructor(networkNode: NetworkNode) {
+    constructor(networkNode: NetworkNode, publishSessionTimeout = 30 * 1000) {
         this.networkNode = networkNode
+        this.publishSessionTimeout = publishSessionTimeout
         this.streams = new Map()
+        this.publishTimeouts = new Map()
+    }
+
+    recordPublish(streamId: string, streamPartition = 0): void {
+        this.networkNode.subscribe(streamId, streamPartition)
+
+        const key = `${streamId}::${streamPartition}`
+        clearTimeout(this.publishTimeouts.get(key)!)
+        this.publishTimeouts.set(key, setTimeout(() => {
+            this.publishTimeouts.delete(key)
+            if (!this.streams.has(key)) {
+                this.networkNode.unsubscribe(streamId, streamPartition)
+            }
+        }, this.publishSessionTimeout))
     }
 
-    subscribe(streamId: string, streamPartition = 0) {
+    subscribe(streamId: string, streamPartition = 0): void {
         const key = `${streamId}::${streamPartition}`
         this.streams.set(key, (this.streams.get(key) || 0) + 1)
 
         this.networkNode.subscribe(streamId, streamPartition)
     }
 
-    unsubscribe(streamId: string, streamPartition = 0) {
+    unsubscribe(streamId: string, streamPartition = 0): void {
         const key = `${streamId}::${streamPartition}`
-        this.streams.set(key, (this.streams.get(key) || 0) - 1)
-
-        if (this.streams.get(key) <= 0) {
-            this.streams.delete(key)
-
-            this.networkNode.unsubscribe(streamId, streamPartition)
+        if (this.streams.has(key)) {
+            if (this.streams.get(key) <= 1) {
+                this.streams.delete(key)
+                if (!this.publishTimeouts.has(key)) {
+                    this.networkNode.unsubscribe(streamId, streamPartition)
+                }
+            } else {
+                this.streams.set(key, this.streams.get(key) - 1)
+            }
```

Comment:
Ugh, lot of nesting here. I strongly prefer early returns than else statements, especially when nested, but just my personal preference:

```js
if (!this.streams.has(key)) { return }

if (this.streams.get(key) > 1) {
    // decrement counter
    this.streams.set(key, this.streams.get(key) - 1)
    return
}

// no more streams
this.streams.delete(key)
// unsubscribe if no pending timeout
if (!this.publishTimeouts.has(key)) {
    this.networkNode.unsubscribe(streamId, streamPartition)
}
```
Very low priority though, IMO better to reduce nesting if possible. Your call.